yw_website/yw_website/apps/website/utils.py

```diff
@@ -43,3 +45,46 @@ def search_and_create_query_set(q):
         return Workflow.objects.filter(query).exclude(version__isnull=True)
     else:
         return Workflow.objects.none()
+
+def get_block_data(run_id):
+    '''
+    This function sets up the block structure so that from a parent program
+    block you may drill down into a child of that program block
+    '''
+    program_blocks = ProgramBlock.objects.filter(run=run_id)
+    data = Data.objects.filter(run=run_id)
+
+    blocks = []
+    data_vals = []
+
+    # grab parent block
+    parent = ProgramBlock()
```

Comment:
I think that it is possible for there to be more than one parent, so there needs to be a list,`parent_blocks`, of your program block objects that gets passed in through the view.

so, ideally you would have something like this:

```
parents = get_direct_descendants(None, program_blocks)

def get_direct_descendants(program_block_id, program_blocks):
    descendants = []
    for child in program_blocks.filter(in_program_block=program_block_id):
        new_child = ProgramBlock()
        new_child.name = child.name
        new_child.programblock_id = child.program_block_id
        new_child.id = child.id
        new_child.direct_descendants = get_direct_descendants(new_child.program_block_id, program_blocks)
        descendants.append(new_child)
    return descendants
```
that should be all the logic you need

Reply:
that's assuming that its possible to filter on None, which im not sure about. There are custom queries for django's queryset objects (the ProgramBlock.objects) called `Q` so if its not possible to do that directly, you can look into these

queryset docs: https://docs.djangoproject.com/en/2.1/ref/models/querysets/
Q docs https://docs.djangoproject.com/en/1.7/ref/models/queries/#q-objects